### PR TITLE
Add message processing depth limit

### PR DESCRIPTION
**Pull request by elyousfi5**
The error_message:
image

### CODE REVIEW

File: src/ostorlab/agent/agent.py [hunk 1]
```diff
@@ -83,6 +87,7 @@ def __init__(
         self.in_selectors = agent_definition.in_selectors
         self.out_selectors = agent_definition.out_selectors
         self.cyclic_processing_limit = agent_settings.cyclic_processing_limit
+        self.processing_depth_limit = agent_settings.processing_depth_limit
```
Comment (3asm) [inline src/ostorlab/agent/agent.py:90]: depth_processing_limit to align naming.

Comment (3asm) [inline src/ostorlab/agent/agent.py:90]: depth_processing_limit to align naming.

File: src/ostorlab/agent/agent.py [hunk 2]
```diff
@@ -203,8 +208,14 @@ def process_message(self, selector: str, message: bytes) -> None:
...
-        except MaximumCyclicProcessReachedError:
-            self.on_max_cyclic_process_reached(object_message)
+        except (
+            MaximumCyclicProcessReachedError,
+            MaximumProcessingDepthReachedError,
+        ) as e:
+            if isinstance(e, MaximumCyclicProcessReachedError):
+                self.on_max_cyclic_process_reached(object_message)
+            elif isinstance(e, MaximumProcessingDepthReachedError):
+                self.on_max_processing_depth_reached(object_message)
```
Comment (3asm) [inline src/ostorlab/agent/agent.py:218]: except X:
   ...
except Y:
   ...

Comment (3asm) [inline src/ostorlab/agent/agent.py:218]: except X:
   ...
except Y:
   ...

File: src/ostorlab/agent/agent.py [hunk 3]
```diff
@@ -234,6 +245,13 @@ def _validate_message(self) -> None:
             ):
                 raise MaximumCyclicProcessReachedError()
 
+        if self.processing_depth_limit is not None and self.processing_depth_limit != 0:
+            if (
+                len(self._control_message.data["control"]["agents"])
+                >= self.processing_depth_limit
+            ):
+                raise MaximumProcessingDepthReachedError()
```
Comment (3asm) [inline src/ostorlab/agent/agent.py:253]: Create a message showing the path from the agents.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: src/ostorlab/agent/agent.py [hunk 4]
```diff
@@ -308,6 +326,15 @@ def on_max_cyclic_process_reached(self, message: agent_message.Message) -> None:
         """
         raise NotImplementedError()
 
+    @abc.abstractmethod
+    def on_max_processing_depth_reached(self, message: agent_message.Message) -> None:
+        """Overridable method triggered on max processing depth reached.
+
+        Returns:
```
Comment (ErebusZ) [inline src/ostorlab/agent/agent.py:333]: No need for this to be in the doc string

**Review by deadly-panda — DISMISSED** [review]
Minor comments, but overall clean implementation, gj!

File: tests/agent/agent_test.py [hunk 5]
```diff
@@ -511,3 +511,149 @@ def run_agent(agent_definition, agent_settings, message, mp_event):
     agent_process.join()
 
     assert mp_event.is_set() is True
+
+
+def testProcessMessage_whenProcessingDepthLimitIsSet_callbackCalled(
```
Comment (benyissa) [inline tests/agent/agent_test.py:516]:
```suggestion
def testProcessMessage_whenProcessingDepthLimitIsReached_callbackCalled(
```

**Review by amine3 — APPROVED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**